draft-ietf-mls-architecture.md

```diff
@@ -317,14 +317,36 @@ members cannot assume that the Member is a member of the group.
 
 The Authentication Service (AS) has to provide two functionalities:
 
-1. authenticate the credentials (i.e. the identity/signature keypair)
-   used in a group
-2. authenticate messages sent in groups given the signature over the
-   message and the sending member's credential
+1. Issue credentials to clients that attest to bindings between identities and
+   signature key pairs
+
+2. Enable a group member to verify that a credential presented by another member
+   is valid
+
+A member with a valid credential authenticates its MLS messages by signing them
+with the private key corresponding to the public key in its credential.
```

Comment:
This does not reflect what we currently have in the spec. I believe we left this up to the AS/application to allow a deniable key exchange OOB, where messages are then signed by the deniable keys. My guess is that most applications will decide to do it as you describe here. At the same time, I think leaving the option for a gap to allow for a deniable key exchange provides some value at almost no cost.

Reply:
I'm not sure how you come to that conclusion from reading the spec.  It looks to me like the spec is very explicit about exactly which keys are used:

[KeyPackage](https://github.com/mlswg/mls-protocol/blob/main/draft-ietf-mls-protocol.md#keypackage-validation)

> Verify that the signature on the KeyPackage is valid using the public key in the LeafNode's credential

[Commit](https://github.com/mlswg/mls-protocol/blob/main/draft-ietf-mls-protocol.md#processing-a-commit)

> In particular, the signature on the enclosing MLSPlaintext MUST verify using the public key for the credential in the leaf_node of the path field.
> External Commits MUST be signed by the new member. In particular, the signature on the enclosing MLSPlaintext MUST verify using the public key for the credential in the leaf_node of the path field.

[GroupInfo](https://github.com/mlswg/mls-protocol/blob/main/draft-ietf-mls-protocol.md#joining-via-welcome-message)

> Verify the signature on the GroupInfo object. The signature input comprises all of the fields in the GroupInfo object except the signature field. The public key and algorithm are taken from the credential in the leaf node of the member with LeafNodeRef signer. 

[MLSPlaintext / MLSCiphertext in general](https://github.com/mlswg/mls-protocol/blob/main/draft-ietf-mls-protocol.md#content-authentication)

> The signature field in an MLSMessageAuth object is computed using the signing private key corresponding to the public key, which was authenticated by the credential at the leaf of the tree indicated by the sender field

Reply:
That said, my understanding of the deniability situation is that the MLS messages are deniable if the AS is deniable.  Suppose the AS authenticates keys deniably, in the sense of not proving to a third party that the relying client didn't make up the authentication proof themself (one could imagine, e.g., an integration with X3DH).  The only thing the MLS signatures prove is that the holder of the private key sent a message.  So if the AS makes the records of who owns which private key falsifiable, then the whole thing is falsifiable.

So there's no need for the layer of ambiguity that you mention.

Reply:
@raphaelrobert @kkohbrok - Can you point me to where in the spec you see this notion that messages can be signed by anything other than the key in the credential?  Because the text cited above says the opposite.

Reply:
In Section 7.1, the spec says:
> The signature field in an MLSMessageAuth object is computed using the signing private key corresponding to the public key, which was authenticated by the credential at the leaf of the tree indicated by the sender field.

I agree that other places in the spec seem to be more explicit about this. In general, it would probably be best if we had only one place that deals with the verification of signatures on an MLSMessage. If we want to keep the option for deniability, we'd have to include this possible level of indirection. It's optional, so people can (and will in most cases) decide to use the signature key in the credential. Also, it really doesn't cost much, except maybe in implementation complexity, but we can put a MAY there, so implementers can choose to allow this or not.

Reply:
This is now resolved with the latest changes in the protocol spec.